src/conn_poll.c

```diff
@@ -141,299 +313,487 @@ int conn_poll_prepare(conn_registry_t *registry, pfds_record_t *pfds, timestamp_
 	*next_timestamp = now + 60000;
 
 	mutex_lock(&registry->mutex);
-	nfds_t size = 1;
+
+	nfds_t needed = 1; // interrupt entry
 	for (int i = 0; i < registry->agents_size; ++i) {
 		juice_agent_t *agent = registry->agents[i];
-		if (!agent) {
+		if (!agent)
 			continue;
-		}
 
 		conn_impl_t *conn_impl = agent->conn_impl;
-		if (!conn_impl ||
-		    (conn_impl->state != CONN_STATE_NEW && conn_impl->state != CONN_STATE_READY)) {
+		if (!conn_impl)
 			continue;
-		}
 
-		size++;
-		if (conn_impl->tcp_sock != INVALID_SOCKET) {
-			size++;
+		if (conn_impl->state != CONN_STATE_NEW && conn_impl->state != CONN_STATE_READY)
+			continue;
+
+		if (conn_impl->state == CONN_STATE_NEW)
+			conn_impl->state = CONN_STATE_READY;
+
+		if (*next_timestamp > conn_impl->next_timestamp)
+			*next_timestamp = conn_impl->next_timestamp;
+
+		++needed; // UDP socket
+		if (conn_impl->tcp_enabled) {
+			if (conn_impl->tcp_listener != INVALID_SOCKET)
+				++needed;
+			needed += conn_impl->tcp_conns_count;
 		}
 	}
 
-	if (pfds->size != size) {
-		struct pollfd *new_pfds = realloc(pfds->pfds, sizeof(struct pollfd) * size);
-		if (!new_pfds) {
-			JLOG_FATAL("Memory allocation for poll file descriptors failed");
+	if (pfds->size != needed) {
+		struct pollfd *new_pfds = realloc(pfds->pfds, sizeof(struct pollfd) * needed);
+		poll_map_entry_t *new_map = realloc(pfds->map, sizeof(poll_map_entry_t) * needed);
+		if (!new_pfds || !new_map) {
+			JLOG_FATAL("Memory allocation for poll data failed");
+			free(new_pfds);
+			free(new_map);
```

Comment:
On failure, `realloc()` returns NULL so freeing is useless. However, if one is successful and not the other, the new array is deallocated but the old pointer for the successful one now points to deallocated memory. It'll be freed on cleanup, leading to a double free memory corruption. You should check results individually.